File: library/src/main/java/io/github/wax911/library/model/request/QueryContainerBuilder.java
package io.github.wax911.library.model.request;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;

import java.util.Map;

/**
 * Created by max on 2018/03/16.
 * Query & Variable builder for graph requests
 */
public class QueryContainerBuilder implements Parcelable {

    private QueryContainer queryContainer;

    public QueryContainerBuilder() {
        queryContainer = new QueryContainer();
    }

    protected QueryContainerBuilder(Parcel in) {
        queryContainer = in.readParcelable(QueryContainer.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(queryContainer, flags);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<QueryContainerBuilder> CREATOR = new Creator<QueryContainerBuilder>() {
        @Override
        public QueryContainerBuilder createFromParcel(Parcel in) {
            return new QueryContainerBuilder(in);
        }

        @Override
        public QueryContainerBuilder[] newArray(int size) {
            return new QueryContainerBuilder[size];
        }
    };

    public QueryContainerBuilder setQuery(String query) {
        this.queryContainer.setQuery(query);
        return this;
    }

    public QueryContainerBuilder putVariable(String key, Object value) {
        queryContainer.putVariable(key, value);
        return this;
    }

    public @Nullable Object getVariable(String key) {
        if(containsVariable(key))
            return queryContainer.getVariables().get(key);
        return null;
    }

    public boolean containsVariable(String key) {
        return queryContainer.containsVariable(key);
    }

    public QueryContainer build() {
        return queryContainer;
    }
}

File: library/src/main/java/io/github/wax911/library/model/request/QueryContainerBuilder.kt
package io.github.wax911.library.model.request

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

/**
 * Created by max on 2018/03/16.
 * Query & Variable builder for graph requests
 */
@Parcelize
class QueryContainerBuilder(private val queryContainer: QueryContainer = QueryContainer()) : Parcelable {

    fun setQuery(query: String?): QueryContainerBuilder {
        this.queryContainer.query = query
        return this
    }

    fun setOperationName(operationName: String?): QueryContainerBuilder{
        this.queryContainer.operationName = operationName
        return this
    }

    fun putVariable(key: String, value: Any?): QueryContainerBuilder {
        queryContainer.putVariable(key, value)
        return this
    }

    fun putVariables(values: Map<String, Any?>): QueryContainerBuilder {
        queryContainer.putVariables(values)
        return this
    }

    fun putExtension(key: String, value: Any?): QueryContainerBuilder {
        queryContainer.putExtension(key, value)
        return this
    }

    fun putPersistedQueryHash(sha256Hash: String, version: Int = 1): QueryContainerBuilder {
        putExtension(persistedQueryExtensionName, PersistedQuery(sha256Hash = sha256Hash, version = version))
        return this
    }

    fun getVariable(key: String): Any? {
        return when {
            containsKey(key) -> queryContainer.variables[key]
            else -> null
        }
    }

    fun containsKey(key: String): Boolean {
        return queryContainer.variables.containsKey(key)
    }

    /**
     * Should only be called by the GraphQLConverter or any other subclasses of it
     * after the query has been added to the request
     * @see io.github.wax911.library.converter.GraphConverter
     */
    fun build(): QueryContainer {
        return queryContainer
    }

    companion object {
        const val persistedQueryExtensionName = "persistedQuery"
    }
}